File: Webfuel.App/ClientApp/src/app/api/project.api.ts
import { EventEmitter, Injectable, inject } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { ApiService, ApiOptions } from '../core/api.service';
import { ActivatedRouteSnapshot, ResolveFn, RouterStateSnapshot } from '@angular/router';
import { IDataSource } from 'shared/common/data-source';
import { UpdateProject, Project, UpdateProjectRequest, UpdateProjectResearcher, UpdateProjectSupportSettings, UnlockProject, CreateTestProject, QueryProject, QueryResult, ReportStep } from './api.types';

@Injectable()
export class ProjectApi implements IDataSource<Project, QueryProject, any, UpdateProject> {
    constructor(private apiService: ApiService) { }
    
    public update (body: UpdateProject, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<UpdateProject, Project>("PUT", "api/project", body, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public updateRequest (body: UpdateProjectRequest, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<UpdateProjectRequest, Project>("PUT", "api/project/request", body, options);
    }
    
    public updateResearcher (body: UpdateProjectResearcher, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<UpdateProjectResearcher, Project>("PUT", "api/project/researcher", body, options);
    }
    
    public updateSupportSettings (body: UpdateProjectSupportSettings, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<UpdateProjectSupportSettings, Project>("PUT", "api/project/support-settings", body, options);
    }
    
    public unlock (body: UnlockProject, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<UnlockProject, Project>("PUT", "api/project/unlock", body, options);
    }
    
    public delete (params: { id: string }, options?: ApiOptions): Observable<any> {
        return this.apiService.request<undefined, any>("DELETE", "api/project/" + params.id + "", undefined, options).pipe(tap(_ => this.changed.emit()));
    }
    
    public createTest (body: CreateTestProject, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<CreateTestProject, Project>("POST", "api/project:create-test", body, options);
    }
    
    public query (body: QueryProject, options?: ApiOptions): Observable<QueryResult<Project>> {
        return this.apiService.request<QueryProject, QueryResult<Project>>("POST", "api/project/query", body, options);
    }
    
    public get (params: { id: string }, options?: ApiOptions): Observable<Project> {
        return this.apiService.request<undefined, Project>("GET", "api/project/" + params.id + "", undefined, options);
    }
    
    public export (body: QueryProject, options?: ApiOptions): Observable<ReportStep> {
        return this.apiService.request<QueryProject, ReportStep>("PUT", "api/project/export", body, options);
    }
    
    changed = new EventEmitter<any>();
    
    // Resolvers
    
    static projectResolver(param: string): ResolveFn<Project> {
        return (route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<Project> => {
            return inject(ProjectApi).get({id: route.paramMap.get(param)! });
        };
    }
}

File: Webfuel.Api/Api/Project/ProjectApi.cs
﻿using MediatR;
using Microsoft.AspNetCore.Mvc;
using Webfuel.Domain;
using Webfuel.Reporting;

namespace Webfuel.App;

[ApiService]
[ApiDataSource]
public static class ProjectApi
{
    public static void RegisterEndpoints(IEndpointRouteBuilder app)
    {
        // Commands

        app.MapPut("api/project", Update)
            .RequireIdentity();

        app.MapPut("api/project/request", UpdateRequest)
            .RequireIdentity();

        app.MapPut("api/project/researcher", UpdateResearcher)
            .RequireIdentity();

        app.MapPut("api/project/support-settings", UpdateSupportSettings)
            .RequireIdentity();

        app.MapPut("api/project/unlock", Unlock)
             .RequireClaim(p => p.CanUnlockProjects);

        app.MapDelete("api/project/{id:guid}", Delete)
            .RequireIdentity();

        app.MapPost("api/project:create-test", CreateTest)
            .RequireClaim(p => p.Developer);

        app.MapPost("api/project:enrich", Enrich)
            .RequireClaim(p => p.Developer);

        // Querys

        app.MapPost("api/project/query", Query)
            .RequireIdentity();

        app.MapGet("api/project/{id:guid}", Get)
            .RequireIdentity();

        app.MapPut("api/project/export", Export)
            .RequireIdentity();
    }

    public static Task<Project> Update([FromBody] UpdateProject command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task<Project> UpdateRequest([FromBody] UpdateProjectRequest command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task<Project> UpdateResearcher([FromBody] UpdateProjectResearcher command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task<Project> UpdateSupportSettings([FromBody] UpdateProjectSupportSettings command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task<Project> Unlock([FromBody] UnlockProject command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task Delete(Guid id, IMediator mediator)
    {
        return mediator.Send(new DeleteProject { Id = id });
    }

    public static Task<Project> CreateTest([FromBody] CreateTestProject command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task Enrich([FromBody] EnrichProject command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static Task<QueryResult<Project>> Query([FromBody] QueryProject command, IMediator mediator)
    {
        return mediator.Send(command);
    }

    public static async Task<Project?> Get(Guid id, IMediator mediator)
    {
        return await mediator.Send(new GetProject { Id = id });
    }

    public static Task<ReportStep> Export([FromBody] QueryProject request, IExportProjectService service)
    {
        return service.InitialiseReport(request);
    }
}
